### DISCUSSION TITLE
Support setting properties with non-primitive values on custom elements

### DISCUSSION THREAD

**Discussion by nwhittaker** [typed-ember/glint#502] [Q&A]
In Ember, given the snippets:

```ts
class MyElement extends HTMLElement {
  myObjProp = { a=0, b=0 }
}
```

```hbs
<my-element myObjProp={{hash a=1 b=2}} />
```

The `MyElement.myObjProp` property gets set to `{ a: 1, b: 2}` without problem. However, Glint fails on the line in the template with:
>Type '{ a: number; b: number; }' is not assignable to type 'AttrValue'.
>Object literal may only specify known properties, and 'a' does not exist in type 'SafeString'.glint:ts(2322)

The error message suggests Glint is treating `myObjProp` as an attribute rather than a property. 

I'm not familiar with Ember's internals, but the behavior I see is that it tries to set the value of a property on the component before falling back to passing the value to the component as an attribute. Any way for Glint to allow setting a component's property with a complex type?

**Comment by dfreeman** [comment]
> it tries to set the value of a property on the component before falling back to passing the value to the component as an attribute

This is true of how Ember handles both web components and native elements, but it's a behavior the framework team has talked for a long time about wanting to deprecate. [This post](https://discuss.emberjs.com/t/what-is-embers-magic-concerning-checkbox-indeterminate/16981/2) has some details on exactly how the heuristic works, and the [strict-mode RFC](https://github.com/emberjs/rfcs/blob/b7e9d794e957e0b3e072aabb9fec6f5ae367c8c5/text/0496-handlebars-strict-mode.md#alternatives) mentions the plan to move to "attribute syntax always means attributes" semantics for templates (though not as part of strict mode).

There's a currently-inactive ["HTML Attribute and Property Rationalization" RFC](https://github.com/emberjs/rfcs/pull/314) that proposes doing away with the prop-vs-attr heuristic behavior, but that wound up stalling out: first because of the desire for general-purpose userspace modifiers, and then because those modifiers don't have access to custom elements early enough in their lifecycle for some use cases. However, there's now [an active RFC to cover _that_](https://github.com/emberjs/rfcs/pull/883), which would give modifiers the ability to act on an element before it's inserted into the DOM.

All of that is to say: currently Glint's restriction on passing non-primitive values as DOM attributes is intentional based on the stated goals of the framework team for future template semantics. In an ideal world, your invocation above would probably look something more like this:
```hbs
<my-element {{props myObjectProp=(hash a=1 b=2)}} />
```
By [adding an entry to the built-in `HTMLElementTagNameMap` interface](https://fireflysemantics.medium.com/adding-the-type-of-a-custom-element-to-the-dom-api-26dd11cfd8fa), you would even get full type safety on that invocation, assuming `props` had a signature something like this:
```ts
interface PropsSignature<E extends Element> {
  Element: E;
  Args: {
    Named: Partial<E>;
  };
}
```

However, if for performance or correctness reasons your component needs all props assigned before its `connectedCallback` is invoked, then something like `{{props}}` probably isn't viable until the modifier timing RFC I mentioned above lands. Additionally, the fact remains that the snippet you posted _does_ work today, even if there's a future where it may not.

Given the above, I can see a few paths we might take in Glint (not all mutually exclusive):
 - Advocate to move [RFC 883](https://github.com/emberjs/rfcs/pull/883) forward as quickly as possible and then guide developers toward using modifiers to handle setting custom element properties
 - Document that developers should use `{{! @glint-expect-error }}` when passing non-primitive values as "attributes" to custom elements
 - Update our internal handling to treat custom elements specially and allow passing arbitrary data via "attributes" on them, with the caveat that you're own your own for type safety.

What are your thoughts, @jamescdavis @chriskrycho?

**Comment by kdagnan** [comment]
Is there any updates to this issue? I see all of the (deeply) linked PRs we're supposedly waiting on, does Glint 2 fix this by some happy chance? (I am still on v1)

**Reply by NullVoxPopuli** [reply]
sorry I have a wip PR that i need to rebase. I've been swamped lately :cry:

**Reply by NullVoxPopuli** [reply]
This is the PR that needs rebasing https://github.com/typed-ember/glint/pull/1016
